### DISCUSSION TITLE
crash in watchOS 8.x when using .task in reducer

### DISCUSSION THREAD

**Discussion by brucejeon** [pointfreeco/swift-composable-architecture#2230] [Q&A]
There is a crash in the WatchOS 8.x environment when returning the EffectTask `.task`, `.run` from the code below. 
WatchOS 9.x works fine.  How should I fix it? (`.task`, `.run` execution line crashes.)
``` swift
var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .refresh:
                return .task { @MainActor in
                    .fetchCouponsResponse(await couponsClient.fetchCoupons())
                }
               or 
                return .run { @MainActor send in
                    send(.fetchCouponsResponse(await couponsClient.fetchCoupons()))
                }
            }
        }
}
```
<img width="1290" alt="스크린샷 2023-06-23 오후 8 25 03" src="https://github.com/pointfreeco/swift-composable-architecture/assets/6380495/de1d52ff-9b06-434b-88f9-21c46905bdb3">

<img width="884" alt="스크린샷 2023-06-26 오후 2 09 16" src="https://github.com/pointfreeco/swift-composable-architecture/assets/6380495/04a0c681-e763-4661-b984-17924059fbe0">

**Comment by mbrandonw** [comment]
Hi @brucejeon, there is not enough information here to diagnose the problem. Can you share a small project that reproduces the problem?

**Reply by brucejeon** [reply]
OK, I'll prepare it and add it

**Comment by brucejeon** [comment]
Hi @mbrandonw share projects that cause the same problem One of the TCA sample projects, `Todo`, has been slightly changed to a watch version, and after running on the watch, you can reproduce the crash crash by selecting the "Add Todo" button -> check comboox.

Crashes occur when `await` is running inside `.run`, `.task` while processing Action, and can only be seen in WatchOS 8.x version of Watch (not Simulator)

It seems difficult to check on the WatchOS 8.x version of Watch, but I hope there is a solution Thank you.


![simulator_screenshot_B725E4B3-BC03-48D5-BBAA-C2D073BEFEBC](https://github.com/pointfreeco/swift-composable-architecture/assets/6380495/2d93c2a9-9f83-485a-8fc9-849b99bcea39)
<img width="731" alt="스크린샷 2023-07-26 오전 8 39 58" src="https://github.com/pointfreeco/swift-composable-architecture/assets/6380495/7cf5dd74-8112-4a12-a615-111e1f59d5f4">
[Todos.zip](https://github.com/pointfreeco/swift-composable-architecture/files/12166387/Todos.zip)

**Reply by mbrandonw** [reply]
Hi @brucejeon, I'm not sure why but your project does not open for me. It is stalled on trying to open in Xcode but never actually opens:

<img width="1072" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/135203/86fbb5f6-e763-4d3b-957f-719110a3ee98">

**Reply by brucejeon** [reply]
Sorry @mbrandonw the project you shared had a TCA dependency on the `Examples folder`
I'm sharing the changed version of SPM again

[Todos.zip](https://github.com/pointfreeco/swift-composable-architecture/files/12168837/Todos.zip)

**Reply by mbrandonw** [reply]
Hi @brucejeon, I'm sorry but I am not able to reproduce this. I have tested on watchOS 8, 9 and 10 and it all seems to work as expected:


https://github.com/pointfreeco/swift-composable-architecture/assets/135203/c94dd0d5-1603-43c3-9aae-9416362836af

**Comment by brucejeon** [comment]
@mbrandonw This crash issue does not occur in the simulator, it only occurs on WatchOS 8.x real devices
So it's hard to test and fix it Please refer to the crash stack of the first image attached when executing `await`

**Reply by mbrandonw** [reply]
I see. Unfortunately I do not have an Apple Watch and so I will not be able to test this anytime soon. I can't help but feel this must be some kind of concurrency bug in Swift. There have been many in older OS versions, and so this just may be another one. I think the fact that it seems to work in watchOS 9 and 10 is also evidence of this. Using Combine may be a workaround until you can drop watchOS 8 support.

**Reply by brucejeon** [reply]
@mbrandonw It's a little disappointing

The problem seems small because the sample I delivered is simple and has no function, but I decided with my team to reflect TCA in the watch app that many users are currently using, and I am checking the user crash report more than I thought

I understand that the scope of TCA support includes WatchOS 8.x, and if you use TCA, isn't it too insignificant to consider the issue that always occurs when you use `await` inside `.task` `.run` as the most basic function used for API calls?

The reason why I hesitated the most when introducing TCA is the third-party framework, so it is difficult to check directly if there is an error inside, but I didn't know I would experience this problem so quickly
In this way, if each app has to deal with exceptions according to the OS version, I think we should seriously consider the rollbeck of TCA application

Please review whether there is a modification method through exception handling within TCA.

**Reply by akbashev** [reply]
@brucejeon 
Even though TCA is a big project with hard work from @mbrandonw, @stephencelis and community—the core and concept of Reducer and Store itself is actually not that complicated (IMHO) and build on top of Apple's frameworks. Debugger from screenshots really suggests it's some issue in async/await. People face issues on specific iOS/macOS versions, so it's possible. I would suggest to double check that first if you have access to watchOS 8.x with some simple ViewModel.

Also, at least to have some crash report would be helpful.

**Reply by mbrandonw** [reply]
> Please review whether there is a modification method through exception handling within TCA.

Hi @brucejeon, I did give an alternative for working around the Swift crash, which is to use Combine and not async/await. That may not be what you want to hear, but unfortunately the fact is that Swift concurrency has lots of known crashes in older iOS/watchOS/etc runtimes (see [here](https://developer.apple.com/forums/thread/729231), [here](https://developer.apple.com/forums/thread/703448), [here](https://forums.swift.org/t/swift-concurrency-back-deploy-issue/53917), [here](https://forums.swift.org/t/weird-crash-with-concurrency/59411), [here](https://forums.swift.org/t/taskgroup-related-crash/56425/7), and a lot more).

As @akbashev mentioned, this project heavily depends on support from the community to keep it healthy. There are hundreds of people that have contributed to the project, both through actual PR's to fix bugs and implement features, but also detailed bug reports to show problems in niche areas that Stephen and I don't work with on a daily basis. watchOS is an example of this, and soon it may be visionOS too.

Perhaps someone can lend their expertise in watchOS to look into this issue further, and maybe there is some simple thing the library can do to work around it. However, my gut tells me that this is just a regular Swift concurrency runtime crash, which you would inevitably run into even in a watchOS 8 project that doesn't use TCA. And so if you really need to support watchOS 8 then you may need to use Combine for a little bit longer.

**Reply by brucejeon** [reply]
Hi @mbrandonw  I understand your position on this issue, and I know I'm the best person to check and fix this issue at this point, so I'll check more on this issue and I'll tell you again if there's any further TCA code modification

And the watch app that is experiencing problems now has already developed api using async/await and converted the api part of the app that is being used without problems into TCA
So I'm sorry to say that it's inevitably run into even in a watchOS 8 project that doesn't use TCA, but it's hard to agree

I came across TCA by developing SwiftUI and reflecting it in our project, learning many things and sharing co-workers,
I want to tell you that I'm always grateful to you for creating TCA

@akbashev The above sample was changed from the simplest sample on the TCA side to operate in the Watch, so it is difficult to make the sample simpler than now
